pkg/abcupdater/version.go

```diff
@@ -127,9 +129,53 @@ func (c *CheckVersionParams) setLocalCachedData(data *LocalVersionData) error {
 	return nil
 }
 
-// CheckAppVersion checks if a newer version of an app is available. Any relevant update info will be
+// CheckAppVersion calls CheckAppVersionSync in a go routine. It returns a closure
+// to be run after program logic which will block until a response is returned
+// or a timeout is hit. Results will be printed to stderr.
+func CheckAppVersion(ctx context.Context, params *CheckVersionParams) func() {
+	return asyncFunctionCall(ctx, func() (string, error) {
+		return CheckAppVersionSync(ctx, params)
+	},
+		os.Stderr)
+}
+
+// asyncFunctionCall handles the async part of CheckAppVersion, but accepts
+// a function and writer as an argument to allow for testability.
+func asyncFunctionCall(ctx context.Context, funcToCall func() (string, error), out io.Writer) func() {
```

Comment:
This function has some complicated requirements on the provided io.Writer that I think need to be documented. It's probably not a big deal since this function isn't part of the public API, but we should still write it down for future maintainers.

The risk here isn't just that writes to this writer will be interleaved with other writes. The risk is a data race (as defined by the [Go Memory Model](https://go.dev/ref/mem)) causing weird program behavior.

One of these must be true:

1. the provided io.Writer is thread-safe
2. the caller does synchronization to ensure that `result()` cannot run concurrently with any other writes to this writer.

Reply:
Changed to a function. Its called during the returned closure, so user has control on when their output function runs.